### ISSUE TITLE
num_shards in SplitGenerator is ignored

### ISSUE THREAD

**Issue by Flamefire** [tensorflow/datasets#1405]
**Short description**
Due to introduction of https://github.com/tensorflow/datasets/pull/650 the `num_shards` parameter in `SplitGenerator` is ignored and instead the TFRecord file writer will calculate an own number of shards in `_get_number_shards`.

**Expected behavior**
`num_shards` is honored if specified

**Additional context**
The number of shards determines the maximum number of parallel instances of models that can be trained in TF. So having it set to some "arbitrary" value without any control makes one depend on this magic value guessing algorithm. Furthermore the dataset_info.json reads inconsistent. E.g.:

```
      "numShards": "1",
      "shardLengths": [
        "128",
        "128",
        "128",
        "128",
        "128",
        "128",
        "128",
        "128",
        "128",
        "128",
        "128",
        "128",
        "128",
        "128",
        "128",
        "128"
      ],
```

For bonus points: Almost every comment/docu on the size of those shards in tensorflow_datasets/core/tfrecords_writer.py is wrong (as in: does not match the actual value): https://github.com/tensorflow/datasets/blob/ef0966d8cd57095c5f3e50bac153ebc53f5dfe34/tensorflow_datasets/core/tfrecords_writer.py#L39 https://github.com/tensorflow/datasets/blob/ef0966d8cd57095c5f3e50bac153ebc53f5dfe34/tensorflow_datasets/core/tfrecords_writer.py#L143 https://github.com/tensorflow/datasets/blob/ef0966d8cd57095c5f3e50bac153ebc53f5dfe34/tensorflow_datasets/core/tfrecords_writer.py#L38

**Comment by Conchylicultor** [comment]
Thanks for reporting, this is indeed a bug and feel inconsistent.

+CC @pierrot0 which added this feature

**Comment by pierrot0** [comment]
Thanks for reaching out and reporting on inconsistent comments.
I'll fix those comments.

numShards was originally here because the pre-processing wasn't figuring it out automatically. So Builder authors were using "random" numbers here. The new way the processing is done, tfrecords_writer figures out a value, which I think is better than most humans tended to specify. 

Old code is still not fully deprecated, so numShards is still around, but I was planning to get rid of it eventually. We could keep it around, if one want to specify the number of shards manually, but I am afraid some authors will just copy paste an existing builder definition...

Alexander, could you tell us more about the feature desire?
Would you have a specific example where [_get_number_shards](https://github.com/tensorflow/datasets/blob/ef0966d8cd57095c5f3e50bac153ebc53f5dfe34/tensorflow_datasets/core/tfrecords_writer.py#L140) would return something which you'd like to override?

Thanks,
Pierre

**Comment by Flamefire** [comment]
> Old code is still not fully deprecated, so numShards is still around

Sure, but at least keep it consistent in the JSON or remove it there completely because shardSizes already contains numShards. At least this is what I'd consider a bug.

> The new way the processing is done, tfrecords_writer figures out a value, which I think is better than most humans tended to specify.

Well it uses some heuristics which I don't see the reasoning explained so I can't tell how good that heuristic is. See below for more.

>  could you tell us more about the feature desire?
Would you have a specific example where _get_number_shards would return something which you'd like to override?

As written above the number of shards has to be bigger than the number of "models in sync" that are run by TF. So this is where the heuristic easily fails: It should be obvious that for a very large model, the data chunks processed by each worker can be small and you get better performance when running more models in parallel. However TFDS is completely unaware of the model size and hence will return the same number of shards for use by small or large models.

So as a model designer/evaluator I may like TFDS to use more shards, a specific (minimum) number or a minimum multiple. E.g. if I know I want to run the model on 128 workers I want a multiple of 128 as the number of shards. If TFDS decides that the input data is small (as can be seen by the comments the definition of "small" even varied during implementing this) then I may only get 32 shards and am unable to run the model with 128 workers.

As a dataset designer I might know, that models tend to be large and overwrite the default to something smaller.

I'd propose the following:
- Allow users to specify the minimum number of shards when loading a dataset. Some conversion could be required if the current number does not match, but that can be offset by doing that beforehand. The implementation will take this number as the minimum and try to use some Power-of-two and/or multiples of that number to split the data
- Allow dataset authors to specify the minimum and maximum size of the shards with the current values as defaults. Iff num_shards is kept, raise an error if it is set or actually use it

The second is mostly optional except for the error if something is requested that is currently silently ignored. But the first is IMO essential: No heuristic can cater for unknown factors so the consumer of a dataset has to have the last say.

**Comment by Conchylicultor** [comment]
I completely agree that we should fix the inconsistency in num_shards, or at least make the two fields consistent.

However, it is by design that the number of shards is independent of the user system. TFDS API guarantee that generating the data on two different system always generate the same data in the same order.
The user should cannot change it at runtime (allowing this would also force to regenerate the dataset everytimes, resulting in multiple copies (each with different numbers of shards)).

Do you have a concrete example of dataset where the heuristic computed by TFDS isn't satisfying enough ? For big dataset, the number of shards should be greater than the number of worker (ex: 1024 shards for imagenet).
For smaller datasets (mnist, cifar,...), the number of shard don't really matter as the dataset fit on memory, and the input pipeline is not the bottleneck.

**Comment by Flamefire** [comment]
> TFDS API guarantee that generating the data on two different system always generate the same data in the same order.

This wouldn't be touched when using more files, would it?

> The user cannot change it at runtime (allowing this would also force to regenerate the dataset everytimes, resulting in multiple copies (each with different numbers of shards)).

It should be designed that it will result in a regeneration from the existing dataset only if the number of shards do not match. If the user specifies the number of shards explicitely and changes it, he pays for the resharding (which BTW is cheaper then redownloading and generating from scratch)   
I don't see this as a problem. There is always only 1 copy with the latest used options (including sharding) and changing them just as when changing the storage dir incurs extra work preparing the data.

> For big dataset, the number of shards should be greater than the number of worker (ex: 1024 shards for imagenet).

Actually this is not complete: The number of shards **must** be greater than the number of workers when using the default TF options: https://www.tensorflow.org/api_docs/python/tf/data/experimental/DistributeOptions#auto_shard_policy

> (ex: 1024 shards for imagenet).

And this is exactly the design confusion I meant: This number is **not** (only) dependent on the dataset but also on the number of workers and (transitively) on the model

> Do you have a concrete example of dataset where the heuristic computed by TFDS isn't satisfying enough ?

Very simple example: MNIST. I can't use multiple workers because there is only 1 shard in the test data.

**Comment by Conchylicultor** [comment]
> This wouldn't be touched when using more files, would it?

It would be. By default shards are read in parallel and interleaved, so changing the shards will change the order in which examples are read.

> The number of shards must be greater than the number of workers when using the default TF options:

The default option are likely not to work because we are using `ds.interleave` instead of a single TFRecordDataset with multiple shards. This is probably something we should improve, to make autosharding works by default (ex: by adding a `ds.shard`, or supporting `input_context` in `ReadConfig` to make `experimental_distribute_datasets_from_function` works).
But this is completely independent feature request from the number of shards. Please open another issue if this is something you want.

> Very simple example: MNIST. I can't use multiple workers because there is only 1 shard in the test data.

People uses MNIST on multiple workers everyday at Google using TFDS, I don't understand what is the issue. As the dataset fit in memory, each worker can train on the full dataset with `ds.shuffle(info.splits['train'].num_examples)`. Otherwise, if you want to make sure each workers see a different part of the data, you can use the subsplit API: https://www.tensorflow.org/datasets/splits (ex: worker 1 read `split='train[50%:]'`, worker 2 read 'train[:50%]')

**Comment by Conchylicultor** [comment]
If I understand correctly, it seems that what you really want is an easier way to set-up distributed tf.data pipelines. I just created a new feature request to track this: https://github.com/tensorflow/datasets/issues/1426

**Comment by Flamefire** [comment]
> It would be. By default shards are read in parallel and interleaved, so changing the shards will change the order in which examples are read.

Ok then maybe document that changing the shard number may change the order?

> The default option are likely not to work because we are using ds.interleave instead of a single TFRecordDataset with multiple shards.

I was following the tutorial at https://www.tensorflow.org/tutorials/distribute/keras. So you are telling me that tutorial is wrong and cannot work? If it was true what you say that TF does not work with the ds.interleave, what is expected to happen instead? Because TF does recognize the underlying number of files used correctly. I.e. the MNIST dataset works in train mode with multiple nodes (~10 shards train data files) but not in test mode (1 shard test data file)

> People uses MNIST on multiple workers everyday at Google using TFDS, I'm not sure I understand the issue.

See the link above: Being a tutorial/docu of TensorFlow this is supposed to work. But it fails once I add `model.evaluate(test_dataset)` (where the dataset is constructed like the train dataset but from the test split) There should be no fiddling with subsplits as TF is supposed to do that. Subsplits are even wrong, because the dataset needs to be shuffled on each epoch before distributing chunks/batches to workers. With your suggestion each worker would always see the same samples which is wrong (as per: training convergence will not work as intended or not at all)   
Additionally in my test it seems as TF is able to distribute the data correctly so each sample is seen once during each epoch even when multiple workers and nodes are used. See https://github.com/tensorflow/datasets/issues/1302

> If I understand correctly, it seems that what you really want is an easier way to set-up distributed tf.data pipelines. 

This is the ultimate goal yes. Feed `tfds.load(...)` into a model under an appropriate strategy and TF should do the shuffling and distribution on an arbitrary number of workers.

**Comment by Conchylicultor** [comment]
Thank you for pointing out to the doc. If indeed, it doesn't works, it is something we should fix. I don't think the fix should involve changing the number of shard though.

Could you paste the full stacktrace that you get when running `model.evaluate(test_dataset)` ideally with a small code snippet to reproduce the issue ?

**Comment by Flamefire** [comment]
Sure:

  ```
File "test_multi_node_distribution.py", line 85, in <module>
    eval_loss, eval_acc = model.evaluate(eval_dataset)
  File "/scratch/ws/s3248973-EasyBuild/easybuild-haswell/software/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training.py", line 930, in evaluate
    use_multiprocessing=use_multiprocessing)
  File "/scratch/ws/s3248973-EasyBuild/easybuild-haswell/software/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training_distributed.py", line 794, in evaluate
    *args, **kwargs)
  File "/scratch/ws/s3248973-EasyBuild/easybuild-haswell/software/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training_distributed.py", line 777, in wrapper
    mode=dc.CoordinatorMode.INDEPENDENT_WORKER)
  File "/scratch/ws/s3248973-EasyBuild/easybuild-haswell/software/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/distribute/distribute_coordinator.py", line 853, in run_distribute_coordinator
    task_id, session_config, rpc_layer)
  File "/scratch/ws/s3248973-EasyBuild/easybuild-haswell/software/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/distribute/distribute_coordinator.py", line 360, in _run_single_worker
    return worker_fn(strategy)
  File "/scratch/ws/s3248973-EasyBuild/easybuild-haswell/software/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training_distributed.py", line 772, in _worker_fn
    return method(model, **kwargs)
  File "/scratch/ws/s3248973-EasyBuild/easybuild-haswell/software/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training_v2.py", line 494, in evaluate
    use_multiprocessing=use_multiprocessing, **kwargs)
  File "/scratch/ws/s3248973-EasyBuild/easybuild-haswell/software/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training_v2.py", line 479, in _model_iteration
    total_epochs=1)
  File "/scratch/ws/s3248973-EasyBuild/easybuild-haswell/software/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training_v2.py", line 128, in run_one_epoch
    batch_outs = execution_function(iterator)
  File "/scratch/ws/s3248973-EasyBuild/easybuild-haswell/software/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training_v2_utils.py", line 98, in execution_function
    distributed_function(input_fn))
  File "/scratch/ws/s3248973-EasyBuild/easybuild-haswell/software/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/eager/def_function.py", line 568, in __call__
    result = self._call(*args, **kwds)
  File "/scratch/ws/s3248973-EasyBuild/easybuild-haswell/software/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/eager/def_function.py", line 638, in _call
    return self._concrete_stateful_fn._filtered_call(canon_args, canon_kwds)  # pylint: disable=protected-access
  File "/scratch/ws/s3248973-EasyBuild/easybuild-haswell/software/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/eager/function.py", line 1611, in _filtered_call
    self.captured_inputs)
  File "/scratch/ws/s3248973-EasyBuild/easybuild-haswell/software/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/eager/function.py", line 1692, in _call_flat
    ctx, args, cancellation_manager=cancellation_manager))
  File "/scratch/ws/s3248973-EasyBuild/easybuild-haswell/software/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/eager/function.py", line 545, in call
    ctx=ctx)
  File "/scratch/ws/s3248973-EasyBuild/easybuild-haswell/software/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/eager/execute.py", line 67, in quick_execute
    six.raise_from(core._status_to_exception(e.code, message), None)
  File "<string>", line 3, in raise_from
tensorflow.python.framework.errors_impl.InvalidArgumentError: 2 root error(s) found.
  (0) Invalid argument:  There aren't enough elements in this dataset for each shard to have at least one element (# elems = 1, # shards = 2). If you are using datasets with distribution strategy, considering setting the auto sharding policy to either DATA or OFF using the `experimental_distribute.auto_shard_policy` optionof `tf.data.Options()`.
	 [[{{node MultiDeviceIteratorGetNextFromShard}}]]
	 [[RemoteCall]]
	 [[IteratorGetNext]]
	 [[allreduce/CollectiveReduce/_12]]
  (1) Invalid argument:  There aren't enough elements in this dataset for each shard to have at least one element (# elems = 1, # shards = 2). If you are using datasets with distribution strategy, considering setting the auto sharding policy to either DATA or OFF using the `experimental_distribute.auto_shard_policy` optionof `tf.data.Options()`.
	 [[{{node MultiDeviceIteratorGetNextFromShard}}]]
	 [[RemoteCall]]
	 [[IteratorGetNext]]
0 successful operations.
```

From:

```
import os
import json
import tensorflow_datasets as tfds
import tensorflow as tf
tfds.disable_progress_bar()

BUFFER_SIZE = 10000
BATCH_SIZE = 64


def make_datasets_unbatched():
    # Scaling MNIST data from (0, 255] to (0., 1.]
    def scale(image, label):
        image = tf.cast(image, tf.float32)
        image /= 255
        return image, label

    datasets, info = tfds.load(name='mnist',
                               with_info=True,
                               as_supervised=True)

    return datasets['train'].map(scale).cache().shuffle(
        BUFFER_SIZE), datasets['test'].map(scale)


def build_and_compile_cnn_model():
    model = tf.keras.Sequential([
        tf.keras.layers.Conv2D(32,
                               3,
                               activation='relu',
                               input_shape=(28, 28, 1)),
        tf.keras.layers.MaxPooling2D(),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(64, activation='relu'),
        tf.keras.layers.Dense(10, activation='softmax')
    ])
    model.compile(loss=tf.keras.losses.sparse_categorical_crossentropy,
                  optimizer=tf.keras.optimizers.SGD(learning_rate=0.001),
                  metrics=['accuracy'])
    return model

strategy = tf.distribute.experimental.MultiWorkerMirroredStrategy()

# Here the batch size scales up by number of workers since
# `tf.data.Dataset.batch` expects the global batch size.
GLOBAL_BATCH_SIZE = BATCH_SIZE * strategy.num_replicas_in_sync
with strategy.scope():
    train_datasets, eval_dataset = (
        ds.batch(GLOBAL_BATCH_SIZE)
        for ds in make_datasets_unbatched())
    model = build_and_compile_cnn_model()

model.evaluate(eval_dataset)
```
